### ISSUE TITLE
Helm chart dependency build fail with private repositories

### ISSUE THREAD

**Issue by romuduck** [argoproj/argo-cd#9863]
Checklist:

* [ x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ x] I've included steps to reproduce the bug.
* [ x] I've pasted the output of `argocd version`.

**Describe the bug**

If a chart dependency comes from a private repository and no associated credentials were found at 1st init of the application, the application will never sync even after adding repo credentials.
The only way to make it sync is to restart the repo-server.
Why:
- at 1st app init, with no matching repo credentials, argocd will perform a `helm repo add xxx https://yyy`
- sync will fail with `chart not found` error
- adding helm repo credentials will trigger `helm repo add zzz https://yyy --username --password`
- and will lead to the following list of helm repos:
    ```
    ❯ helm repo list
    NAME         	URL                                                                                      
    xxx	        https://yyy    < no creds
    zzz      	https://yyy    < with creds
   ```

- sync will still fail as will attempt downloading chart from 1st matching repo in the list, which has no creds.

**To Reproduce**
Try to sync a helm chart with a dependency chart coming from a private repo. It should fail.
Add a helm repo credential.
Try to sync again the application.
It still fail.

**Expected behavior**
It should not fail.
Maybe a proposal would be, when adding a new helm repository credentials/template:
-   check if repo or template matches one/several in `helm repo list`
-   remove them
the next sync should add again the now missing repo with right credentials.

**Version**

```shell
argocd: v2.3.0+fe42780.dirty
  BuildDate: 2022-03-07T05:19:08Z
  GitCommit: fe427802293b090f43f91f5839393174df6c3b3a
  GitTreeState: dirty
  GoVersion: go1.17.8
  Compiler: gc
  Platform: darwin/amd64
```

**Comment by mmisztal1980** [comment]
I'm experiencing a similar issue with ArgoCD 4.8.3 and working with Azure DevOps and Azure Container Registry:

HELM Repos are defined using ArgoCD's HELM values file:

```yaml
configs:
  repositories:
    devops-acr-helm-repo:
      type: helm
      url: myacr.azurecr.io/charts/kedacore
      name: kedacore
      username: <service-principal-with-acr-pull-app-id>
      password: <service-principal-with-acr-pull-password>
      enableOCI: "true"
```

HELM & GIT Repos are visible in the **Repositories** view:
![image](https://user-images.githubusercontent.com/1357057/178306883-b2c0bd6a-743f-4312-929b-7daa7a4857a2.png)

Apps are modelled as HELM subcharts in their respective repositories:

**Chart.yaml**
```yaml
apiVersion: v2
name: keda-subchart
type: application
version: 1.0.0
appVersion: "1.0.0"
dependencies:
- name: keda
  version: 2.6.0
  repository: "@kedacore"
```

Attempt to sync the app fails with:
![image](https://user-images.githubusercontent.com/1357057/178307404-3f6c8216-154f-4e56-a941-7858c336edde.png)

Quick inspection of the HELM repos in the repo server pods yields:
```bash
k -n argocd exec -it argo-argocd-repo-server-54848f994d-74mpr -- bash
Defaulted container "repo-server" out of: repo-server, copyutil (init)
argocd@argo-argocd-repo-server-54848f994d-74mpr:~$ helm repo list
Error: no repositories to show
```


The problem appears to be similar to what @romuduck is describing

**Comment by mmisztal1980** [comment]
If I understand the [code](https://github.com/argoproj/argo-cd/blob/f34687bd80a8977052f2ce9cd145c8f0265d9b59/util/helm/helm.go#L73) correctly (I'm not a GO developer) : 
- if the repository is an OCI repository, argo is going to:
  - perform `helm registry login`
  - attempt to execute `helm dependencies build`
  - perform `helm registry logout`
  - ensure the `h.cmd.IsHelmOci` is set

I'm guessing this is a purely HELM issue similar to https://github.com/helm/helm/issues/10608

**Comment by blakepettersson** [comment]
This should be fixed with #11327, let me know if that's not the case 😄

**Comment by lado936** [comment]
same issue with 
Argo CD
v2.13.1+af54ef8
